### ISSUE TITLE
Hello from xiaomiRobot

### ISSUE THREAD

**Issue by Roborock-OpenSource** [marcelrv/XiaomiRobotVacuumProtocol#11]
hi Marcel,
very great job,
thank you for your interest in our robot .
please be free to contact me when you have question about xiaomiRobot(Roborock).

**Comment by marcelrv** [comment]
Thank you @Roborock-OpenSource 

I think the most prominent question we have is how to determine the token of a device in a non-hacked way. It would be great if there is a clear way we can logon to the cloud and receive the token without the need to downgrade the android app or need to put custom firmware on the robot.

**Comment by Roborock-OpenSource** [comment]
I will discuss with our team, my email is opensource@roborock.com ,  can you send e email about this ?

**Comment by marcelrv** [comment]
yes, please find a more elaborate mail in your email box.
I'll also write some more thoughts that would make OS implementations easier.
e.g. easy endpoint to retrieve the mapdata from the device to allow OS apps to incorporate the cleaning map

**Comment by Luc3as** [comment]
Hello, would be possible to add support for newer protocol which uses Xiaowa E20/E35 vacuums ?

**Comment by marcelrv** [comment]
@Roborock-OpenSource did you receive my email?
Can you feedback if this was useful and some response if this is something you can support this or not, or maybe still in discussion?
thx

**Comment by ciB89** [comment]
I'm also curious how the new commands are to set up virtual no-go zones and walls - I would like to create a virtual no-go zone for my carpets when mobbing the floor and delete it after cleaning again.

**Comment by Luc3as** [comment]
> I'm also curious how the new commands are to set up virtual no-go zones and walls - I would like to create a virtual no-go zone for my carpets when mobbing the floor and delete it after cleaning again.

Do you already have this functionality in APP  ? I did read somewhere there could be some big update from Mi at the end of the this year with this functionality.

**Comment by ciB89** [comment]
> > I'm also curious how the new commands are to set up virtual no-go zones and walls - I would like to create a virtual no-go zone for my carpets when mobbing the floor and delete it after cleaning again.
> 
> Do you already have this functionality in APP ? I did read somewhere there could be some big update from Mi at the end of the this year with this functionality.

Yes, it's in my app. The functionality has been on chinese servers for a couple of weeks now, but it got unlocked today in Europe too! It also enables you to save the map.

**Comment by dugite-code** [comment]
FYI if you don't have the new options showing @FeatureExpert has found a [way to enable this with root access]( https://github.com/dgiese/dustcloud/issues/140#issuecomment-438998917) over on the dustcloud github

**Comment by ciB89** [comment]
any updates?

**Comment by nuc134r** [comment]
You can obitain token without root or hacking the robot using patched Mi Home app. It also broadcasts token in it's own WiFi network before it is paired. 

The protocol workflow and tokens are well described [in this repo](https://github.com/skysilver-lab/php-miio). It is in Russian but I can volunteer to translate it to English someday.

Patched version of Mi Home app which allows to get token is available on it's developer's [blog](http://www.kapiba.ru/2017/11/mi-home.html) also in Russian. 

Use Chrome's "Translate to English" from context menu. :)

**Comment by dugite-code** [comment]
@ciB89 have a look at https://github.com/marcelrv/XiaomiRobotVacuumProtocol/issues/15 some preliminary discoveries regarding the virtual walls are there.

**Comment by andreashaertel** [comment]
Hi, I have written a C++ code to send and receive UDP packets to a S5 Max that is uninitialized. I am able to receive the device id and the token. However, now I am preparing a packet to send WiFi credentials or a packet to send a "Start" command and the robot is not reacting. 
First: is there any interest in this direction? 
Second: anyone able to help me getting this running? 
My plan is to connect the robot to a WiFi without internet connection and to send commands via the program. 

I followed the description on https://github.com/OpenMiHome/mihome-binary-protocol/blob/master/doc/PROTOCOL.md

**Comment by marcelrv** [comment]
@andreashaertel Similar features are available in the Openhab & python-miio.
Is there something specific that you are tying to achieve that is not already there?

**Comment by andreashaertel** [comment]
Thank you, I had no idea about the python-miio: it works great. Now I like to learn about the internal coordinates of the device, but this seems to be complicated (without connecting it to the cloud or hacking its firmware).
